File: caddisfly-app/app/src/main/java/org/akvo/caddisfly/sensor/cuvette/ResultFragment.java
/*
 * Copyright (C) Stichting Akvo (Akvo Foundation)
 *
 * This file is part of Akvo Caddisfly.
 *
 * Akvo Caddisfly is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Akvo Caddisfly is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Akvo Caddisfly. If not, see <http://www.gnu.org/licenses/>.
 */

package org.akvo.caddisfly.sensor.cuvette;


import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;

import com.google.gson.Gson;

import org.akvo.caddisfly.R;
import org.akvo.caddisfly.databinding.FragmentResult2Binding;
import org.akvo.caddisfly.model.QualityGuide;
import org.akvo.caddisfly.model.Result;
import org.akvo.caddisfly.model.Standard;
import org.akvo.caddisfly.model.TestInfo;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.List;
import java.util.Objects;

import static org.akvo.caddisfly.common.ConstantKey.IS_INTERNAL;
import static org.akvo.caddisfly.common.ConstantKey.TEST_INFO;

public class ResultFragment extends Fragment {

    private FragmentResult2Binding b;

    /**
     * Get the instance.
     */
    public static ResultFragment newInstance(TestInfo testInfo, boolean isInternal) {
        ResultFragment fragment = new ResultFragment();
        Bundle args = new Bundle();
        args.putParcelable(TEST_INFO, testInfo);
        args.putBoolean(IS_INTERNAL, isInternal);
        fragment.setArguments(args);
        return fragment;
    }

    private static String readStringFromResource(Context ctx, int resourceID) {
        StringBuilder contents = new StringBuilder();
        String sep = System.getProperty("line.separator");

        try {
            InputStream is = ctx.getResources().openRawResource(resourceID);

            try (BufferedReader input = new BufferedReader(new InputStreamReader(is), 1024 * 8)) {
                String line;
                while ((line = input.readLine()) != null) {
                    contents.append(line);
                    contents.append(sep);
                }
            }
        } catch (IOException ex) {
            return null;
        }

        return contents.toString();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        b = DataBindingUtil.inflate(inflater,
                R.layout.fragment_result_2, container, false);
        View view = b.getRoot();

        if (getArguments() != null) {
            TestInfo testInfo = getArguments().getParcelable(TEST_INFO);
            if (testInfo != null) {
                setInfo(testInfo);
            }
        }

        return view;
    }

    public void setInfo(TestInfo testInfo) {
        Result result = testInfo.getResults().get(0);

        if (result != null) {
            String json = readStringFromResource(Objects.requireNonNull(getActivity()),
                    R.raw.quality_guide_ind);

            b.safeInfoLayout.setVisibility(View.VISIBLE);

            List<Standard> standards = new Gson().fromJson(json, QualityGuide.class).getStandards();
            for (Standard standard : standards) {
                if (standard.getUuid() != null && standard.getUuid().equalsIgnoreCase(testInfo.getUuid())) {
                    if (standard.getMax() != null && result.getResultValue() > standard.getMax()) {
                        b.unsafeInfoLayout.setVisibility(View.VISIBLE);
                        b.safeInfoLayout.setVisibility(View.GONE);
                    }

                    if (standard.getMin() != null && standard.getMin() < result.getResultValue()) {
                        b.unsafeInfoLayout.setVisibility(View.VISIBLE);
                        b.safeInfoLayout.setVisibility(View.GONE);
                    }
                    break;
                }
            }

            b.textResult.setText(result.getResult());
            b.textTitle.setText(testInfo.getName());
            b.textDilution.setText(getResources().getQuantityString(R.plurals.dilutions,
                    testInfo.getDilution(), testInfo.getDilution()));
            b.textUnit.setText(result.getUnit());

            if (testInfo.getDilution() == testInfo.getMaxDilution()) {
                b.textDilutionInfo.setVisibility(View.GONE);
            } else if (result.highLevelsFound()) {
                b.textDilutionInfo.setVisibility(View.VISIBLE);
            } else {
                b.textDilutionInfo.setVisibility(View.GONE);
            }
        }
    }
}

File: caddisfly-app/app/src/main/java/org/akvo/caddisfly/sensor/chamber/ResultFragment.kt
package org.akvo.caddisfly.sensor.chamber

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import kotlinx.android.synthetic.main.fragment_result.*
import org.akvo.caddisfly.R
import org.akvo.caddisfly.common.ConstantKey.IS_INTERNAL
import org.akvo.caddisfly.common.ConstantKey.TEST_INFO
import org.akvo.caddisfly.databinding.FragmentResultBinding
import org.akvo.caddisfly.model.TestInfo
import org.akvo.caddisfly.util.toLocalString
import java.util.*

class ResultFragment : Fragment() {
    private var mListener: OnDilutionSelectedListener? = null
    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val b: FragmentResultBinding = DataBindingUtil.inflate(
            inflater,
            R.layout.fragment_result, container, false
        )

        return b.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        requireActivity().setTitle(R.string.result)

        if (arguments != null) {
            val testInfo: TestInfo? = requireArguments().getParcelable(TEST_INFO)
            if (testInfo != null) {
                val result = testInfo.results!![0]
                textResult.text = result.result
                textTitle.text = testInfo.name!!.toLocalString()
                textDilution.text = resources.getQuantityString(
                    R.plurals.dilutions,
                    testInfo.dilution, testInfo.dilution
                )
                textUnit.text = result.unit
                if (testInfo.dilutions.size > 1) {
                    when {
                        testInfo.dilution > 1 -> {
                            buttonDilution1.visibility = View.GONE
                            buttonCustomDilution.visibility = View.GONE
                            high_level_txt.visibility = View.GONE
                        }
                        result.highLevelsFound() -> {
                            buttonDilution1.visibility = View.VISIBLE
                            high_level_txt.visibility = View.VISIBLE
                            val dilution = testInfo.dilutions[1]
                            buttonDilution1.text = String.format(
                                Locale.getDefault(),
                                getString(R.string.times_dilution), dilution
                            )
                            buttonDilution1.setOnClickListener {
                                mListener!!.onDilutionSelected(
                                    dilution
                                )
                            }
                            if (testInfo.dilutions.size > 2) {
                                buttonCustomDilution.visibility = View.VISIBLE
                                buttonCustomDilution.setOnClickListener { showCustomDilutionDialog() }
                            } else {
                                buttonCustomDilution.visibility = View.GONE
                            }
                        }
                        else -> {
                            buttonDilution1.visibility = View.GONE
                            high_level_txt.visibility = View.GONE
                        }
                    }
                }
            }
        }
    }

    private fun showCustomDilutionDialog() {
        if (activity != null) {
            val ft = requireActivity().supportFragmentManager.beginTransaction()
            val editCustomDilution = EditCustomDilution.newInstance()
            editCustomDilution.show(ft, "editCustomDilution")
        }
    }

    interface OnDilutionSelectedListener {
        fun onDilutionSelected(dilution: Int)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        mListener = if (context is OnDilutionSelectedListener) {
            context
        } else {
            throw IllegalArgumentException(
                context.toString()
                        + " must implement OnDilutionSelectedListener"
            )
        }
    }

    companion object {
        /**
         * Get the instance.
         */
        fun newInstance(testInfo: TestInfo?, isInternal: Boolean): ResultFragment {
            val fragment = ResultFragment()
            val args = Bundle()
            args.putParcelable(TEST_INFO, testInfo)
            args.putBoolean(IS_INTERNAL, isInternal)
            fragment.arguments = args
            return fragment
        }
    }
}